sol-anchor/idl.go

```diff
@@ -0,0 +1,612 @@
+	}
+
+	if f.IsOption() {
+		if f.Option.IsSimple() {
+			return PrintSimple(*f.Option.Simple, fieldName, variableName, true)
+		}
+
+		if f.Option.IsDefined() {
+			return PrintDefined(*f.Option.Defined, fieldName, variableName, types, false, true)
+		}
+	}
+
+	return ""
+}
+
+/*
+A field could be of type:
+    - simple (e.g. "string")
+    - defined
+    - array
+    - vec
+	- optional (simple, defined or vec)
+*/
+func unmarshalDefined(data []byte) string {
+	var definedType struct {
+		Defined string `json:"defined"`
+	}
+	if err := json.Unmarshal(data, &definedType); err == nil && definedType.Defined != "" {
+		return definedType.Defined
+	}
+
+	return ""
+}
+
+func unmarshalSimple(data []byte) string {
+	var simpleType string
+	if err := json.Unmarshal(data, &simpleType); err == nil && simpleType != "" {
+		return simpleType
+	}
+
+	return ""
+}
+
+/*
+Return types:
+	bool: isDefined (true/false)
+	string: type (simple/defined)
+*/
+func unmarshalVec(data []byte) (bool, string) {
+	// Try simple
+	var vecSimpleType struct {
+		Vec string `json:"vec"`
+	}
+	err := json.Unmarshal(data, &vecSimpleType)
+	if err == nil {
+		return false, vecSimpleType.Vec
+	}
+
+	// Try defined
+	type DefinedType struct {
+		Defined string `json:"defined"`
+	}
+	var vecDefinedType struct {
+		Vec DefinedType `json:"vec"`
+	}
+	err = json.Unmarshal(data, &vecDefinedType)
+	//fmt.Printf("%s", string(vecDefinedType.Vec.Defined))
+	if err == nil {
+		return true, vecDefinedType.Vec.Defined
+	}
+
+	return false, ""
+}
+
+/*
+Return types:
+	- option type (simple, defined or vec)
+	- type
+*/
+func unmarshalOption(data []byte) (string, string) {
+	// Try simple
+	var optionSimpleType struct {
+		Option string `json:"option"`
+	}
+	if err := json.Unmarshal(data, &optionSimpleType); err == nil && optionSimpleType.Option != "" {
+		return "simple", optionSimpleType.Option
+	}
+
+	// Try defined
+	type DefinedType struct {
+		Defined string `json:"defined"`
+	}
+	var optionDefinedType struct {
+		Option DefinedType `json:"option"`
+	}
+	if err := json.Unmarshal(data, &optionDefinedType); err == nil && optionDefinedType.Option.Defined != "" {
+		return "defined", optionDefinedType.Option.Defined
+	}
+
+	// Try vec simple
+	type VecTypeSimple struct {
+		Vec string `json:"vec"`
+	}
+	var optionVecSimpleType struct {
+		Option VecTypeSimple `json:"option"`
+	}
+	if err := json.Unmarshal(data, &optionVecSimpleType); err == nil && optionVecSimpleType.Option.Vec != "" {
+		return "vecSimple", optionVecSimpleType.Option.Vec
+	}
+
+	// Try vec defined
+	type VecTypeDefined struct {
+		Vec DefinedType `json:"vec"`
+	}
+	var optionVecDefinedType struct {
+		Option VecTypeDefined `json:"option"`
+	}
+	if err := json.Unmarshal(data, &optionVecDefinedType); err == nil && optionVecDefinedType.Option.Vec.Defined != "" {
+		return "vecDefined", optionVecDefinedType.Option.Vec.Defined
+	}
+
+	// TODO: Try array?
+
+	return "", ""
+}
+
```

Comment:
The "fields" of an Event or an Instruction could of different types: simple, defined, vec, array or optional.

For example:

Simple (e.g. string, u64...)
```json
{
  type: "string"
}
```

Defined (user-defined structs)
```json
{
  type: {
    "defined": "MyCustomType"
  }
}
```

So because the that JSON part could be of different types, we have to try which one matches (not sure if there is a better way of doing it)